### PR TITLE
Return to 1a433d45, bump version

### PR DESCRIPTION
**Pull request by relatko**

### CODE REVIEW

File: app/src/coin.h [hunk 1]
```diff
@@ -37,35 +35,14 @@ extern "C" {
...
-    return (path.data[0] == HDPATH_0_DEFAULT && path.data[1] == HDPATH_1_DEFAULT);
-}
-
-__Z_INLINE bool path_is_testnet(const flow_path_t path) { //or emulatornet
-    return (path.data[0] == HDPATH_0_TESTNET && path.data[1] == HDPATH_1_TESTNET);
-}
-
-__Z_INLINE bool path_is_empty(const flow_path_t path) {
-    return path.data[0] == 0 && path.data[1] == 0;
-}
```
Comment (tarakby) [inline app/src/coin.h:55]: I think we can keep these functions

File: app/src/account.c [hunk 2]
```diff
@@ -208,10 +204,13 @@ zxerr_t slot_parseSlot(uint8_t *buffer, uint16_t bufferLen) {
         return zxerr_out_of_bounds;
     }
 
-    //We copy account and path. Note that path is copied in app_main.c:extractHDPath (for sign transaction call).
     MEMCPY(&tmp_slot, buffer + 1, sizeof(account_slot_t));
 
-    if (!path_is_mainnet(tmp_slot.path) && !path_is_testnet(tmp_slot.path) && !path_is_empty(tmp_slot.path)) {
+    const bool mainnet = tmp_slot.path.data[0] == HDPATH_0_DEFAULT && tmp_slot.path.data[1] == HDPATH_1_DEFAULT;
+    const bool testnet = tmp_slot.path.data[0] == HDPATH_0_TESTNET && tmp_slot.path.data[1] == HDPATH_1_TESTNET;
+    const bool empty = tmp_slot.path.data[0] == 0 && tmp_slot.path.data[1] == 0;
```
Comment (tarakby) [inline app/src/account.c:211]: we could keep the functions `path_is_mainnet` etc

File: app/src/account.c [hunk 3]
```diff
@@ -224,27 +223,6 @@ zxerr_t slot_parseSlot(uint8_t *buffer, uint16_t bufferLen) {
...
-        }
-        if (path_is_testnet(tmp_slot.path)) {
-            if (!validateChainAddress(CODEWORD_TESTNET, account) && 
-                    !validateChainAddress(CODEWORD_EMULATORNET, account)) {
-                zemu_log_stack("invalid account");
-                return zxerr_out_of_bounds;
-            }
-        }
-    }
-
```
Comment (tarakby) [inline app/src/account.c:247]: I think any information that ledger gets from the outside and represents an account address should be validates using `vaildateChainAddress`. Is there a reason for deleting the validation here?

File: app/src/apdu_handler.c [hunk 4]
```diff
@@ -31,42 +31,12 @@
...
-
-    MEMCPY(&hdPath, slot.path.data, sizeof(hdPath)); 
-
-    //We fill account and pubkey
-    MEMCPY(G_io_apdu_buffer, &slot.account, sizeof(slot.account));
+    extractHDPath(rx, OFFSET_DATA);
 
-    *tx = sizeof(slot.account);
-    *tx += app_fill_pubkey(G_io_apdu_buffer + sizeof(slot.account), IO_APDU_BUFFER_SIZE - sizeof(slot.account)); 
+    uint8_t requireConfirmation = G_io_apdu_buffer[OFFSET_P1];
```
Comment (tarakby) [inline app/src/apdu_handler.c:36]:
```suggestion
    uint8_t displayAndConfirm = G_io_apdu_buffer[OFFSET_P1];
```

File: app/src/common/actions.h [hunk 5]
```diff
@@ -46,14 +45,18 @@ __Z_INLINE void app_reject() {
     io_exchange(CHANNEL_APDU | IO_RETURN_AFTER_TX, 2);
 }
 
-__Z_INLINE uint8_t app_fill_pubkey(unsigned char *buffer, uint16_t buffer_len) {
-    zxerr_t err = crypto_extractPublicKey(&hdPath, buffer, buffer_len);
+__Z_INLINE uint8_t app_fill_address() {
```
Comment (tarakby) [inline app/src/common/actions.h:48]:
```suggestion
__Z_INLINE uint8_t app_fill_pubkey() {
```

File: app/src/common/actions.h [hunk 6]
```diff
@@ -46,14 +45,18 @@ __Z_INLINE void app_reject() {
...
+    // Put data directly in the apdu buffer
+    MEMZERO(G_io_apdu_buffer, IO_APDU_BUFFER_SIZE);
 
-    if (err != zxerr_ok) {
+    action_addr_len = 0;
+    zxerr_t err = crypto_fillAddress(hdPath, G_io_apdu_buffer, IO_APDU_BUFFER_SIZE - 2, &action_addr_len);
```
Comment (tarakby) [inline app/src/common/actions.h:53]:
```suggestion
    zxerr_t err = crypto_fill_pubkey(hdPath, G_io_apdu_buffer, IO_APDU_BUFFER_SIZE - 2, &action_addr_len);
```

File: app/src/apdu_handler.c [hunk 7]
```diff
@@ -31,42 +31,12 @@
...
 
-    //Display or return the buffer    
-    if (displayAndConfirm) {
-        //we set the returning length for app_reply_address
-        action_addr_len = *tx;
+    if (requireConfirmation) {
+        app_fill_address();
```
Comment (tarakby) [inline app/src/apdu_handler.c:39]:
```suggestion
        app_fill_pubkey();
```

File: app/src/apdu_handler.c [hunk 8]
```diff
@@ -75,6 +45,7 @@ __Z_INLINE void handleGetPubkey(volatile uint32_t *flags, volatile uint32_t *tx,
         return;
     }
 
+    *tx = app_fill_address();
```
Comment (tarakby) [inline app/src/apdu_handler.c:48]:
```suggestion
    *tx = app_fill_pubkey();
```

File: app/src/common/app_main.c [hunk 9]
```diff
@@ -95,10 +95,15 @@ void extractHDPath(uint32_t rx, uint32_t offset) {
...
+    MEMCPY(hdPath, G_io_apdu_buffer + offset, sizeof(uint32_t) * HDPATH_LEN_DEFAULT);
 
-    if (!path_is_mainnet(hdPath) && !path_is_testnet(hdPath)) {
+    const bool mainnet = hdPath[0] == HDPATH_0_DEFAULT &&
+                         hdPath[1] == HDPATH_1_DEFAULT;
+
+    const bool testnet = hdPath[0] == HDPATH_0_TESTNET &&
+                         hdPath[1] == HDPATH_1_TESTNET;
+
+    if (!mainnet && !testnet) {
```
Comment (tarakby) [inline app/src/common/app_main.c:106]: same for `path_is_mainnet` and `path_is_testnet`

File: app/src/crypto.c [hunk 10]
```diff
@@ -18,7 +18,7 @@
 #include "coin.h"
 #include "zxmacros.h"
 
-flow_path_t hdPath;
+uint32_t hdPath[HDPATH_LEN_DEFAULT];
```
Comment (tarakby) [inline app/src/crypto.c:21]: using `flow_path_t` was also a good refactor

File: app/src/crypto.c [hunk 11]
```diff
@@ -62,22 +62,23 @@ __Z_INLINE cx_curve_t get_cx_curve(const uint32_t path[HDPATH_LEN_DEFAULT]) {
...
+    cx_curve_t curve = get_cx_curve(path);
     if (curve!=CX_CURVE_SECP256K1 && curve!=CX_CURVE_SECP256R1 ) {
         zemu_log_stack("extractPublicKey: invalid_crypto_settings");
         return zxerr_invalid_crypto_settings;
     }
 
-    if (pubKeyLen < PUBLIC_KEY_LEN) {
+    const uint32_t domainSize = 32;
+    const uint32_t pkSize = 1 + 2 * domainSize;
+    if (pubKeyLen < pkSize) {
```
Comment (tarakby) [inline app/src/crypto.c:77]: we could keep using `PUBLIC_KEY_LEN` that is equal to `65`
```suggestion
    if (pubKeyLen < PUBLIC_KEY_LEN) {
```

File: app/src/crypto.c [hunk 12]
```diff
@@ -226,4 +227,28 @@ zxerr_t crypto_sign(const uint32_t path[HDPATH_LEN_DEFAULT], const uint8_t *mess
     return zxerr_ok;
 }
 
+typedef struct {
+    uint8_t publicKey[SECP256K1_PK_LEN];
+    char addrStr[SECP256K1_PK_LEN*2];
+    uint8_t padding[4];
+} __attribute__((packed)) answer_t;
+
+zxerr_t crypto_fillAddress(const uint32_t path[HDPATH_LEN_DEFAULT], uint8_t *buffer, uint16_t buffer_len, uint16_t *addrLen) {
```
Comment (tarakby) [inline app/src/crypto.c:236]:
```suggestion
zxerr_t crypto_fill_pubkey(const uint32_t path[HDPATH_LEN_DEFAULT], uint8_t *buffer, uint16_t buffer_len, uint16_t *addrLen) {
```

**Review by tarakby — COMMENTED** [review]
Thanks for reverting the changes. 
I left some comments below. Some good refactors and cleanups were made as part of the recent changes, and I think it makes sense to not revert them.

**Comment by relatko** [comment]
@tarakby Thanks for identifying the refactors we want to keep. I prefer to do it in another commit. The idea is to make the review process trivial for ledger guys (they review against 0.9.10, and they find only 4 new transactions with a straightforward, and already reviewed code. This should make the review process faster. But of course, in the end we want these refactors.

**Comment by tarakby** [comment]
@relatko sounds good, that works for me!

**Review by Wolog2021 — APPROVED** [review]

**Pull request merged**